### DISCUSSION TITLE
getStaticPaths should have access to the query params

### DISCUSSION THREAD

**Discussion by Vince-vegas** [vercel/next.js#17616] [Ideas]
Next.js should put a feature on getStaticPaths to have access on Params for more dynamic routes

e.g 

When you have a site like netflix the **ID**'s of their movies are large, in your getStaticPaths you should have access to the params so you can dynamically fetch the query id on routes

**Comment by lucasbasquerotto** [comment]
I think that this would be very important, and it's even expected now that fallback is possible, being possible to generate static pages dynamically based on an id/slug. I think that query params should be accepted too, and render the same static page if the query params are the same.

For example, in my site I have pages with urls like `/projects` and `/projects?page=2`, and I don't think I should change the urls of the site to something like `/projects/2` to be able to use static generation, especially considering that NextJS is already able to handle dynamic ids in the path.

In this case it would be important to be able to specify which parameters I want to use (like `page`), because parameters like the one when you access a link with Facebook (`fbid`) shouldn't generate another html (because they are not relevant to the page).

Or an alternative would be to make NextJS use the page at `pages/projects/[page].js` when the user accesses the url `/projects?page=2`, in this case I would handle internally to use the page mentioned before through the url `pages/projects/2`, **but the browser would still show `/projects?page=2`**. Is it possible? That is, is it possible to convert the route `/projects?page=[page]` to `pages/projects/[page]` in the application?

**Reply by agusterodin** [reply]
Did you end up finding an elegant solution?

**Reply by lucasbasquerotto** [reply]
@agusterodin In cases that the page only has 1 parameter (or very few parameters) you can use a [rewrite](https://nextjs.org/docs/api-reference/next.config.js/rewrites) rule from `/page` (with query parameter `foo`) to `page/:bar`, where `bar` is the value of the `foo` parameter extracted as a regex in the rule (for example, `value: '(?<bar>*)'`, as can be seen similar examples in the docs about rewrites). Unfortunately, I don't know of a general solution, tough. There's a [feature request asking for that](https://github.com/vercel/next.js/discussions/17269), but I don't see an official reply about if (and when) it will be implemented.

**Reply by agusterodin** [reply]
Thanks! For whatever reason this doesn't seem to be working (the question mark really screws it up). Am I doing something really dumb? I want to match `/register?plan=pro`

```
async rewrites() {
    return [
      {
        source: '/register\\?plan=pro',
        destination: '/register/pro'
      }
    ]
  }
```

**Reply by lucasbasquerotto** [reply]
@agusterodin You should use the `has` property to match the query, something like:

```yaml
{
    source: '/register',
    has: [
        {
            type: 'query',
            key: 'plan',
            value: '(?<plan>.+)',
        },
    ],
    destination: '/register/:plan',
}
```

_(I haven't tested the above code, but should be something similar to it. The regex `.+` is to match any value, but you could change to something like `(?<plan>basic|pro)`, for example.)_

**Reply by agusterodin** [reply]
Hell yeah, works perfectly

**Reply by lhguerra** [reply]
with this approach wouldn't we end up with two possibilities for the same route? (/register?plan=foo and /register/foo)

**Reply by lucasbasquerotto** [reply]
@lhguerra This approach is for cases in which you use only the query parameters, so the page shouldn't be allowed directly with `/register/foo` (if it is, I don't see much benefit of also having the query parameter, so you could use only the path parameter). 

If you want access with the query parameter (only), in `next.config` you can deny the path to `/register/foo`, only allowing `/register` (which also accept query parameters).

**Reply by lhguerra** [reply]
thank's, this deny feature is just what I need :D

**Reply by lucasbasquerotto** [reply]
@lhguerra In my setup I place all internal pages in a specific directory, and block acess to this specific directory (it can only be accessed using rewrites). This way I don't need to deny each page specifically.

**Reply by abbou2** [reply]
@lucasbasquerotto please if I have multi parameters for example a filter like this : /register?page=2&name=test but the order of those params can change

**Reply by lucasbasquerotto** [reply]
@abbou2 My approach would be only for single parameters, or if there's a way to use a catch-all for parameters, and keep in mind that NextJS does not support query parameters yet. I specified a better approach at https://github.com/vercel/next.js/discussions/17269#discussioncomment-106569, to handle multiple parameters, but it should be implemented by the NextJS team. That said, it doesn't seem that vercel will make changes to `getStaticProps` now that they are focusing on using the `app` directory.
